### ISSUE TITLE
Array indexing logic treats all arrays are one-dimensional

### ISSUE THREAD

**Issue by andrioni** [MaterializeInc/materialize#9815]
### What version of Materialize are you using?

v0.14.1-dev (851347e86)

### How did you install Materialize?

Built from source

### What was the issue?

Right now, the indexing operation on arrays treats all arrays as if they were row-major one-dimensional arrays, so `array[[1,2],[3,4],[5,6]]` is accessed as if it were `array[1,2,3,4,5,6]`, while Postgres always enforces access by each dimension separately. A couple of examples to illustrate:

### Materialize
```sql
materialize=> select (array[[1,2],[3,4],[5,6]])[1];
 ?column?
----------
        1
(1 row)

materialize=> select (array[[1,2],[3,4],[5,6]])[5];
 ?column?
----------
        5
(1 row)

materialize=> select (array[[1,2],[3,4],[5,6]])[1][1];
ERROR:  cannot subscript type integer
materialize=> select (array[[1,2],[3,4],[5,6]])[3][1];
ERROR:  cannot subscript type integer
```

### Postgres
```sql
alessandro=# select (array[[1,2],[3,4],[5,6]])[1];
 array
-------

(1 row)

alessandro=# select (array[[1,2],[3,4],[5,6]])[5];
 array
-------

(1 row)

alessandro=# select (array[[1,2],[3,4],[5,6]])[1][1];
 array
-------
     1
(1 row)

alessandro=# select (array[[1,2],[3,4],[5,6]])[3][1];
 array
-------
     5
(1 row)
```

**Comment by sploiselle** [comment]
We don't really want to be in the business of supporting arrays any more than necessary ([docs](https://materialize.com/docs/sql/types/array/)); is there a specific tool that uses this feature that we want to unblock?

**Comment by andrioni** [comment]
Ah, not really, I filed this one mostly so that I wouldn't forget about it, but I also consider it pretty low priority overall, especially since it only affects multidimensional arrays, which are rarely used.

**Comment by sploiselle** [comment]
Turns out this is necessary for `int2vector[]`, which wants to use `Datum::Array` as its internal representation

**Comment by benesch** [comment]
Fixed in #10705.
